File: LeetCode2021-2022/ts/17.letterCombinationsOfAPhone.ts
type dict = {[key:string]:string[]};
const backTrack = (digits:string, d2l: dict,combina:string[], x:number,
				   p:string):void =>{
	const n:number = digits.length;
	if(x === n){
		combina.push(p);
		return;
	}

	for(let alpha of d2l[digits[x]]){
		backTrack(digits, d2l, combina, x+1, p + alpha);
	}
	return;
}
const letterCombinations = (digits:string):string[] =>{
	const d2l:dict = {
        "2":  ["a","b","c"],
        "3":  ["d", "e","f"],
        "4": ["g","h","i"],
        "5": ["j","k","l"],
        "6":["m","n","o"],
        "7":["p","q","r","s"],
        "8":["t","u","v"],
         "9":["w","x","y","z"]
    };
	let combina:string[] = [];
	backTrack(digits,d2l, combina, 0, "");
	return combina;
}

File: LeetCode2021-2022/python/2023/17.letterCombinationsOfAPhone.py
class Solution:
    from collections import defaultdict
    def letterCombinations(self, digits: str) -> List[str]:
        n = len(digits)
        if n == 0:
            return []
        table = defaultdict(list)
        table['2'] = ['a', 'b', 'c']
        table['3'] = ['d', 'e', 'f']        
        table['4'] = ['g', 'h', 'i']        
        table['5'] = ['j', 'k', 'l']        
        table['6'] = ['m', 'n', 'o']        
        table['7'] = ['p', 'q', 'r', 's']        
        table['8'] = ['t', 'u', 'v']        
        table['9'] = ['w', 'x', 'y', 'z']
        
        combina = [];
        def backTrack(x:int, p:str) -> None:
            if x == n:
                combina.append(p)
                return
            
            for alpha in table[digits[x]]:
                backTrack(x+1, p + alpha)
            return
        backTrack(0, "")
        return combina
